sled-agent/src/metrics.rs

```diff
@@ -4,124 +4,415 @@
+/// this interval, we stop attempting to collect its data.
+const TRANSIENT_LINK_EXPIRATION_INTERVAL: Duration = Duration::from_secs(60);
 
 /// The maximum Dropshot request size for the metrics server.
 const METRIC_REQUEST_MAX_SIZE: usize = 10 * 1024 * 1024;
 
+/// Size of the queue used to send messages to the metrics task.
+const QUEUE_SIZE: usize = 64;
+
 /// An error during sled-agent metric production.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     #[error("Kstat-based metric failure")]
     Kstat(#[source] KstatError),
 
-    #[error("Failed to insert metric producer into registry")]
-    Registry(#[source] MetricsError),
+    #[error("Failed to start metric producer server")]
+    ProducerServer(#[source] oximeter_producer::Error),
+}
 
-    #[error("Failed to fetch hostname")]
-    Hostname(#[source] std::io::Error),
+/// Messages sent to the sled-agent metrics collection task.
+///
+/// The sled-agent publish metrics to Oximeter, including statistics about
+/// datalinks. This metrics task runs in the background, and code that creates
+/// or deletes objects can notify this task to start / stop tracking statistics
+/// for them.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(test, allow(dead_code))]
+pub(crate) enum Message {
+    /// Start tracking the named physical link.
+    ///
+    /// This is only use on startup, to track the underlays.
+    TrackPhysical { zone_name: String, name: String },
+    /// Track the named VNIC.
+    TrackVnic { zone_name: String, name: String },
+    /// Stop tracking the named VNIC.
+    UntrackVnic { name: String },
+    /// Track the named OPTE port.
+    TrackOptePort { zone_name: String, name: String },
+    /// Stop tracking the named OPTE port.
+    UntrackOptePort { name: String },
+    // TODO-completeness: We will probably want to track other kinds of
+    // statistics here too. For example, we could send messages when a zone is
+    // created / destroyed to track zonestats; we might also want to support
+    // passing an explicit `oximeter::Producer` type in, so that other code
+    // could attach their own producers.
+}
 
-    #[error("Non-UTF8 hostname")]
-    NonUtf8Hostname,
+/// Helper to define kinds of tracked links.
+struct LinkKind;
 
-    #[error("Missing NULL byte in hostname")]
-    HostnameMissingNull,
+impl LinkKind {
+    const PHYSICAL: &'static str = "physical";
+    const VNIC: &'static str = "vnic";
+    const OPTE: &'static str = "opte";
+}
 
-    #[error("Failed to start metric producer server")]
-    ProducerServer(#[source] oximeter_producer::Error),
+/// The main task used to collect and publish sled-agent metrics.
+async fn metrics_task(
+    sled_identifiers: SledIdentifiers,
+    kstat_sampler: KstatSampler,
+    _server: ProducerServer,
+    log: Logger,
+    mut rx: mpsc::Receiver<Message>,
+) {
+    let mut tracked_links: BTreeMap<String, TargetId> = BTreeMap::new();
```

Comment:
does ordering matter of the keys here? Just wondering if with the un-tracking/tracking in place, would a hashmap with arbitrary keys work better (though I know this is a change separate from the PR itself)?

Reply:
No, ordering doesn't matter, I just tend to reach for a B-tree first out of habit. We can make this a hashmap if that's preferable, I don't have a strong opinion.

Reply:
yeah. I mean, it seems to be the more reasonable one for arbitrary lookups, etc, but no biggie of course for this specific case.

Reply:
Switched to hash-map in f9d4b5c3f